lib/porky_lib/unencrypted/file_service.rb

```diff
@@ -0,0 +1,53 @@
+# frozen_string_literal: true
+
+require 'singleton'
+require 'porky_lib/file_service_helper'
+
+class Unencrypted::FileService
+  include Singleton
+  include FileServiceHelper
+
+  class FileServiceError < StandardError; end
+  class FileSizeTooLargeError < StandardError; end
+
+  def read(bucket_name, file_key, options = {})
+    tempfile = Tempfile.new
+
+    begin
+      object = s3.bucket(bucket_name).object(file_key)
+      raise FileSizeTooLargeError, "File size is larger than maximum allowed size of #{max_file_size}" if object.content_length > max_size
+
+      object.download_file(tempfile.path, options)
+    rescue Aws::Errors::ServiceError => e
+      raise FileServiceError, "Attempt to download a file from S3 failed.\n#{e.message}"
+    end
+
+    tempfile.read
+  end
+
+  def write(file, bucket_name, options = {})
+    raise FileServiceError, 'Invalid input. One or more input values is nil' if input_invalid?(file, bucket_name)
+    raise FileSizeTooLargeError, "File size is larger than maximum allowed size of #{max_file_size}" if file_size_invalid?(file)
+
+    data = file_data(file)
+    file_key = options.key?(:directory) ? "#{options[:directory]}/#{SecureRandom.uuid}" : SecureRandom.uuid
+
+    tempfile = write_tempfile(data, file_key)
```

Comment:
Why do we need to write this tempfile? Why not use the file directly to upload to s3?

Reply:
We could write it directly, what this does is just write the file and renames it to a uuid, where as the file passed would contain the date.

Also, the method can take as param for file either the file or the data (so there could be no file already created). Do you think we should only be allowing to pass a file?

Reply:
There is potential for significant performance hit if we copy the file when not necessary. You can do the File.file?(file) check here and decide if you need to read the data or not:

```tempfile = File.file?(file) ? file : write_tempfile(file_data(file), file_key)```

(also remove line 32)

Reply:
Ok sounds good!